squint.c

```diff
@@ -0,0 +1,1611 @@
+         int *funcEnd;
+         int numReturn = 0;
+         int *retAddr = 0;
+         ++scan;
+         while (scan < end) {
+            if (*scan == 0xe92d4800 && !is_const(scan)) { // push {fp, lr}
+               break;
+            }
+            else if (*scan == 0xe8bd8800 && !is_const(scan)) { // pop {fp, pc}
+               ++numReturn;
+               retAddr = scan;
+            }
+            ++scan;
+         }
+         --scan;
+         funcEnd = scan /* retAddr */;
+
+         /******************************************/
+         /***   convert stack VM to frame VM     ***/
+         /******************************************/
+
+         simplify_branch(funcBegin, retAddr);
+         apply_peepholes1(funcBegin, retAddr);
+         apply_peepholes2(tmpbuf, funcBegin, retAddr);
+         apply_peepholes1(funcBegin, retAddr);
+         apply_peepholes3(tmpbuf, funcBegin, retAddr);
+         create_pushpop_map(tmpbuf, funcBegin, retAddr);
+         apply_peepholes4(funcBegin, retAddr);
+
+         /******************************************/
+         /***  convert frame VM to register VM   ***/
+         /******************************************/
+
+         rename_register1(funcBegin, retAddr);
+
+         rename_nop(funcBegin, retAddr);
+         funcEnd = relocate_nop(funcBegin, funcEnd, 0);
+      }
+      else {
+         ++scan;
+      }
+   }
+
+   destroy_const_map();
+   free(tmpbuf);
+}
+
+/**********************************************************/
+/******** read exe, optimize, write exe *******************/
+/**********************************************************/
+
+int main(int argc, char *argv[])
+{
+   int fd;
+   int offset, length;
+   int *mem;
+
+   if (argc != 4) {
+      printf("Use: %s <amacc executable> <text start> <length>\n", argv[0]);
+      exit(-1);
+   }
+
+   offset = strtol(argv[2], 0, 16);
+   length = strtol(argv[3], 0, 16);
+
+   if (!(offset > 0 && length > 0)) {
+       printf("last two arguments must be hex file offset and length\n");
+       exit(-1);
+   }
+
+   mem = (int *) malloc(length);
+
+   if ((fd = open(argv[1], O_RDWR)) < 0) {
+      printf("could not open file %s\n", argv[1]);
+      exit(-1);
+   }
+
+   if (lseek(fd, (off_t) offset, SEEK_SET) != offset) {
+      printf("could not seek to offset %x in file %s.\n", offset, argv[1]);
+      exit(-1);
+   }
+
+   if (read(fd, mem, (size_t)length) != length) {
+      printf("could not read %x bytes from file %s.\n", length, argv[1]);
+      exit(-1);
+   }
+
+   peephole_opt(mem, mem + length/4);
+
+   if (lseek(fd, (off_t) offset, SEEK_SET) != offset) {
+      printf("could not seek to offset %x in file %s.\n", offset, argv[1]);
+      exit(-1);
+   }
+
+   if (write(fd, mem, (size_t)length) != length) {
+      printf("error occured attempting to write file %s\n", argv[1]);
+      exit(-1);
+   }
+
+   close(fd);
+
+   free(mem);
+
+   printf("Please remember the amacc '-p' compiler flag is a prerequisite.\n");
```

Comment:
Maybe write a real check that checks whether the binary is compiled with the `-p` flag?  The condition can be:

```
int is_peephole_func_prologue(int *code) {
   if (*code++ != 0xe92d4800 ||  // push {fp, lr}
       *code++ != 0xe28db000) {  // add  fp, sp, #0
      return 0;
   }

   if ((*code & 0xffffff00) != 0xe24dd000) {  // sub  sp, sp, #(tmp * 4)
      if (*code != NOP) {
         return 0;
      }
      num_nop--;
      code++;
   }

   while (num_nop > 0) {
      if (*code++ != NOP) {
         return 0;
      }
   }
   return 1;
}
```